File: src/main/java/ru/org/linux/group/GroupListDao.java
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.group;

import com.google.common.collect.ImmutableList;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.support.rowset.SqlRowSet;
import org.springframework.stereotype.Repository;
import ru.org.linux.section.Section;
import ru.org.linux.topic.Topic;
import ru.org.linux.topic.TopicPermissionService;
import ru.org.linux.topic.TopicTagService;
import ru.org.linux.tracker.TrackerFilterEnum;
import ru.org.linux.user.User;
import ru.org.linux.user.UserNotFoundException;
import ru.org.linux.user.UserService;
import ru.org.linux.util.StringUtil;

import javax.sql.DataSource;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Repository
public class GroupListDao {
  private final NamedParameterJdbcTemplate jdbcTemplate;

  public GroupListDao(UserService userService, TopicTagService topicTagService, DataSource ds) {
    this.userService = userService;
    this.topicTagService = topicTagService;
    jdbcTemplate = new NamedParameterJdbcTemplate(ds);
  }

  private final UserService userService;

  private final TopicTagService topicTagService;

  private static final String queryTrackerMain =
      "WITH topics AS (" +
        "SELECT topics.*, groups.id as gid, sections.moderate as smod, groups.title AS gtitle, urlname, section FROM topics " +
        "JOIN groups ON topics.groupid=groups.id JOIN sections ON sections.id = groups.section " +
        "WHERE not draft" +
        "%s" + // deleted
        "%s" + /* user!=null ? queryPartIgnored*/
        "%s" + // queryAuthorFilter
        "%s" + // queryPartTagIgnored
        "%s" + // noUncommited
        "%s" + // partFilter
        "%s" + // innerSortLimit
      ") SELECT * FROM (SELECT DISTINCT ON(id) * FROM (SELECT " +
        "t.userid as author, " +
        "t.id, lastmod, " +
        "t.stat1 AS stat1, " +
        "gid, " +
        "gtitle, " +
        "t.title AS title, " +
        "comments.id as cid, " +
        "comments.userid AS last_comment_by, " +
        "t.resolved as resolved," +
        "section," +
        "urlname," +
        "comments.postdate as comment_postdate, " +
        "smod, " +
        "t.moderate, " +
        "t.sticky, " +
        "t.postdate as topic_postdate, " +
        "t.deleted, " +
        "t.postscore as topic_postscore " +
      "FROM topics AS t JOIN comments ON (t.id=comments.topic) " +
      "WHERE t.postscore IS DISTINCT FROM " + TopicPermissionService.POSTSCORE_HIDE_COMMENTS + " " +
        "AND comments.id=(SELECT id FROM comments WHERE NOT deleted AND comments.topic=t.id " +
          "%s" + /* user!=null ? queryCommentIgnored */
          "%s" + // queryAuthorFilter
          "ORDER BY postdate DESC LIMIT 1) " +
        "%s" + // commentInterval
     "UNION ALL " +
       "SELECT " +
          "t.userid as author, " +
          "t.id, lastmod,  " +
          "t.stat1 AS stat1, " +
          "gid, " +
          "gtitle, " +
          "t.title AS title, " +
          "0, " + /*cid*/
          "0, " + /*last_comment_by*/
          "t.resolved as resolved," +
          "section," +
          "urlname," +
          "postdate as comment_postdate, " +
          "smod, " +
          "t.moderate, " +
          "t.sticky, " +
          "t.postdate as topic_postdate, " +
          "t.deleted, " +
          "t.postscore as topic_postscore " +
       "FROM topics AS t " +
       "%s " + // WHERE topicInterval
     ") as tracker ORDER BY id, comment_postdate desc) tracker " +
     "%s"; // outerSortLimit

  private static final String queryPartCommentIgnored =
          " AND not exists (select ignored from ignore_list where userid=:userid intersect select get_branch_authors(comments.id)) ";
  private static final String queryPartIgnored =
          " AND userid NOT IN (select ignored from ignore_list where userid=:userid) ";
  private static final String queryPartTagIgnored = " AND topics.id NOT IN (select tags.msgid from tags, user_tags "
    + "where tags.tagid=user_tags.tag_id and user_tags.is_favorite = false and user_id=:userid " +
          "except select tags.msgid from tags, user_tags where " +
          "tags.tagid=user_tags.tag_id and user_tags.is_favorite = true and user_id=:userid) ";
  private static final String queryPartNoTalks = " AND not topics.groupid = 8404 ";
  private static final String queryPartTech = " AND not topics.groupid in (8404, 4068, 9326, 19405) AND section=2 ";
  private static final String queryPartMain = " AND not topics.groupid in (8404, 4068, 9326, 19405) ";

  private static final String noUncommited = " AND (topics.moderate or NOT sections.moderate) ";

  public List<TopicsListItem> getGroupTrackerTopics(int groupid, Optional<User> currentUser, int topics, int offset,
                                                    int messagesInPage, Optional<Integer> tagId) {

    String dateFilter = ">CURRENT_TIMESTAMP-'6 month'::interval ";

    String partFilter = " AND topics.groupid = " + groupid + " ";

    String tagFilter = tagId.map(t -> " AND topics.id IN (SELECT msgid FROM tags WHERE tagid="+t+") ").orElse("");

    return load(partFilter + tagFilter + " AND lastmod"+dateFilter, "", currentUser,
            topics, offset, messagesInPage, "comment_postdate",
            "AND comments.postdate"+dateFilter,
            "t.postdate"+dateFilter, false, false);
  }

  public List<TopicsListItem> getGroupListTopics(int groupid, Optional<User> currentUser, int topics, int offset,
                                                 int messagesInPage, boolean showIgnored, boolean showDeleted,
                                                 Optional<Integer> year, Optional<Integer> month, Optional<Integer> tagId) {
    String dateInterval;

    if (year.isPresent()) {
      dateInterval="postdate>='" + year.get() + '-' + month.get() + "-01'::timestamp AND " +
              "(postdate<'" + year.get() + '-' + month.get() + "-01'::timestamp+'1 month'::interval)";
    } else  {
      dateInterval = "postdate>CURRENT_TIMESTAMP-'6 month'::interval ";
    }

    String partFilter = " AND topics.groupid = " + groupid + " AND NOT topics.sticky AND " + dateInterval;

    String tagFilter = tagId.map(t -> " AND topics.id IN (SELECT msgid FROM tags WHERE tagid="+t+") ").orElse("");

    return load(partFilter + tagFilter, "", currentUser,
            topics, offset, messagesInPage,
            "topic_postdate", "", "", showIgnored, showDeleted);
  }

  public List<TopicsListItem> getSectionListTopics(Section section, Optional<User> currentUser, int topics, int offset,
                                                   int messagesInPage, Integer tagId) {
    String partFilter = " AND section = " + section.getId();

    String tagFilter = " AND topics.id IN (SELECT msgid FROM tags WHERE tagid="+tagId+") ";

    return load(partFilter + tagFilter, "", currentUser,
            topics, offset, messagesInPage,
            "topic_postdate", "", "", false, false);
  }

  public List<TopicsListItem> getGroupStickyTopics(Group group, int messagesInPage, Optional<Integer> tagId) {
    String partFilter = " AND topics.groupid = " + group.getId() + " AND topics.sticky ";

    String tagFilter = tagId.map(t -> " AND topics.id IN (SELECT msgid FROM tags WHERE tagid="+t+") ").orElse("");

    return load(partFilter + tagFilter, "", Optional.empty(),
            100, 0, messagesInPage,
            "topic_postdate", "", "", true, false);
  }

  public List<TopicsListItem> getTrackerTopics(TrackerFilterEnum filter, Optional<User> currentUser,
                                               int topics, int offset, int messagesInPage) {
    String partFilter = switch (filter) {
      case NOTALKS -> queryPartNoTalks;
      case MAIN -> queryPartMain;
      case TECH -> queryPartTech;
      default -> "";
    };

    String userFilter = switch (filter) {
      case SCORE50 -> " AND userid IN (SELECT id FROM users WHERE score<50) ";
      case SCORE100 -> " AND userid IN (SELECT id FROM users WHERE score<100) ";
      default -> "";
    };

    String dateFilter = ">CURRENT_TIMESTAMP-'7 days'::interval ";

    return load(partFilter +" AND lastmod"+dateFilter, userFilter, currentUser, topics, offset, messagesInPage, "comment_postdate",
            "AND comments.postdate"+dateFilter,
            "t.postdate"+dateFilter, false, false);
  }

  private List<TopicsListItem> load(String partFilter, String authorFilter, Optional<User> currentUserOpt,
                                    int topics, int offset, final int messagesInPage, String orderColumn,
                                    String commentInterval, String topicInterval, boolean showIgnored,
                                    boolean showDeleted) {
    String innerSortLimit;
    String outerSortLimit;

    // если сортируем по топику, то можно заранее отобрать нужные топики,
    // до получения даты последнего комментария
    if (orderColumn.equals("topic_postdate")) {
      innerSortLimit = "ORDER BY postdate DESC LIMIT " + topics + " OFFSET " + offset;
      outerSortLimit = "ORDER BY topic_postdate DESC";
    } else {
      innerSortLimit = "";
      outerSortLimit = "ORDER BY " + orderColumn + " DESC LIMIT " + topics + " OFFSET " + offset;
    }

    User currentUser = currentUserOpt.orElse(null);

    MapSqlParameterSource parameter = new MapSqlParameterSource();

    String partIgnored;
    String commentIgnored;
    String tagIgnored;

    if (currentUser != null && !showIgnored) {
      commentIgnored = queryPartCommentIgnored;
      partIgnored = queryPartIgnored;
      tagIgnored = queryPartTagIgnored;
      parameter.addValue("userid", currentUser.getId());
    } else {
      partIgnored = "";
      commentIgnored = "";
      tagIgnored = "";
    }

    boolean showUncommited = currentUser!=null && (currentUser.isModerator() || currentUser.canCorrect());

    String partUncommited = showUncommited ? "" : noUncommited;
    String partDeleted = showDeleted?"":" AND NOT deleted ";

    String query;

    query = String.format(
            // topics CTE
            queryTrackerMain, partDeleted, partIgnored, authorFilter, tagIgnored, partUncommited, partFilter, innerSortLimit,
            // comments part
            commentIgnored, authorFilter, commentInterval,
            // topics part
            topicInterval.isEmpty()?"":("WHERE " + topicInterval),
            // order
            outerSortLimit);

    SqlRowSet resultSet = jdbcTemplate.queryForRowSet(query, parameter);

    List<TopicsListItem> res = new ArrayList<>(topics);
    
    while (resultSet.next()) {
      User author = userService.getUserCached(resultSet.getInt("author"));
      int msgid = resultSet.getInt("id");
      Timestamp lastmod = resultSet.getTimestamp("lastmod");
      int stat1 = resultSet.getInt("stat1");
      int groupId = resultSet.getInt("gid");
      String groupTitle = resultSet.getString("gtitle");
      String title = StringUtil.makeTitle(resultSet.getString("title"));
      int cid = resultSet.getInt("cid");
      User lastCommentBy;
      try {
        int id = resultSet.getInt("last_comment_by");

        if (id != 0) {
          lastCommentBy = userService.getUserCached(id);
        } else {
          lastCommentBy = null;
        }
      } catch (UserNotFoundException e) {
        throw new RuntimeException(e);
      }
      boolean resolved = resultSet.getBoolean("resolved");
      int section = resultSet.getInt("section");
      String groupUrlName = resultSet.getString("urlname");
      Timestamp postdate = resultSet.getTimestamp("comment_postdate");
      boolean sticky = resultSet.getBoolean("sticky");
      boolean uncommited = resultSet.getBoolean("smod") && !resultSet.getBoolean("moderate");
      int pages = Topic.pageCount(stat1, messagesInPage);

      ImmutableList<String> tags;

      tags = topicTagService.getTagsForTitle(msgid);

      int topicPostscore = (resultSet.getObject("topic_postscore") == null)
              ? TopicPermissionService.POSTSCORE_UNRESTRICTED
              : resultSet.getInt("topic_postscore");

      res.add(new TopicsListItem(author, msgid, lastmod, stat1,
              groupId, groupTitle, title, cid, lastCommentBy, resolved,
              section, groupUrlName, postdate, uncommited, pages, tags, resultSet.getBoolean("deleted"),
              sticky, topicPostscore));
    }
    
    return res;
  }
}

File: src/main/scala/ru/org/linux/group/GroupListDao.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.group

import org.springframework.stereotype.Repository
import ru.org.linux.auth.AnySession
import ru.org.linux.scalikejdbc.SpringDB
import ru.org.linux.section.Section
import ru.org.linux.section.SectionController.NonTech
import ru.org.linux.topic.TopicPermissionService
import ru.org.linux.tracker.TrackerFilterEnum
import ru.org.linux.util.StringUtil
import ru.org.linux.warning.WarningService.TopicMaxWarnings
import scalikejdbc.*

object GroupListDao:
  private def queryPartCommentIgnored(userId: Int): SQLSyntax =
    sqls"AND NOT EXISTS (SELECT ignored FROM ignore_list WHERE userid=${userId} INTERSECT SELECT get_branch_authors(comments.id))"

  private def queryPartIgnored(userId: Int): SQLSyntax =
    sqls"AND userid NOT IN (SELECT ignored FROM ignore_list WHERE userid=${userId})"

  private def queryPartTagIgnored(userId: Int): SQLSyntax =
    sqls"AND topics.id NOT IN (SELECT tags.msgid FROM tags, user_tags WHERE tags.tagid=user_tags.tag_id AND user_tags.is_favorite=false AND user_id=${userId} EXCEPT SELECT tags.msgid FROM tags, user_tags WHERE tags.tagid=user_tags.tag_id AND user_tags.is_favorite=true AND user_id=${userId})"

  private val NoUncommited: SQLSyntax = sqls"AND (topics.moderate OR NOT sections.moderate)"

@Repository
class GroupListDao(springDB: SpringDB):

  def getGroupTrackerTopics(groupid: Int, offset: Int, tagId: Option[Int])(using
      session: AnySession): collection.Seq[TopicsListItem] =
    val partFilter = SQLSyntax.join(
      Seq(
        sqls"AND topics.groupid = ${groupid}",
        tagId.map(t => sqls"AND topics.id IN (SELECT msgid FROM tags WHERE tagid=${t})").getOrElse(SQLSyntax.empty),
        sqls"AND lastmod > CURRENT_TIMESTAMP - '6 month'::interval"
      ),
      sqls" "
    )

    load(
      partFilter = partFilter,
      topics = session.profile.topics,
      offset = offset,
      sortByTopic = false,
      commentInterval = sqls"AND comments.postdate > CURRENT_TIMESTAMP - '6 month'::interval",
      topicInterval = sqls"t.postdate > CURRENT_TIMESTAMP - '6 month'::interval",
      showIgnored = false,
      showDeleted = false,
      showUncommited = false
    )

  def getGroupListTopics(
      groupid: Int,
      offset: Int,
      showIgnored: Boolean,
      showDeleted: Boolean,
      yearMonth: Option[(Int, Int)],
      tagId: Option[Int])(using session: AnySession): collection.Seq[TopicsListItem] =
    val dateInterval: SQLSyntax = yearMonth
      .map { (year, month) =>
        sqls"postdate >= make_date(${year}, ${month}, 1)::timestamp AND postdate < make_date(${year}, ${month}, 1)::timestamp + '1 month'::interval"
      }
      .getOrElse(sqls"postdate > CURRENT_TIMESTAMP - '6 month'::interval")

    val partFilter = SQLSyntax.join(
      Seq(
        sqls"AND topics.groupid = ${groupid}",
        sqls"AND NOT topics.sticky",
        sqls"AND ${dateInterval}",
        tagId.map(t => sqls"AND topics.id IN (SELECT msgid FROM tags WHERE tagid=${t})").getOrElse(SQLSyntax.empty)
      ),
      sqls" "
    )

    load(
      partFilter = partFilter,
      topics = session.profile.topics,
      offset = offset,
      sortByTopic = true,
      commentInterval = SQLSyntax.empty,
      topicInterval = SQLSyntax.empty,
      showIgnored = showIgnored,
      showDeleted = showDeleted,
      showUncommited = false
    )

  def getSectionListTopics(section: Section, offset: Int, tagId: Int)(using
      session: AnySession): collection.Seq[TopicsListItem] =
    val partFilter = SQLSyntax.join(
      Seq(sqls"AND section = ${section.id}", sqls"AND topics.id IN (SELECT msgid FROM tags WHERE tagid=${tagId})"),
      sqls" ")

    load(
      partFilter = partFilter,
      topics = session.profile.topics,
      offset = offset,
      sortByTopic = true,
      commentInterval = SQLSyntax.empty,
      topicInterval = SQLSyntax.empty,
      showIgnored = false,
      showDeleted = false,
      showUncommited = false
    )

  def getGroupStickyTopics(group: Group, tagId: Option[Int])(using
      session: AnySession): collection.Seq[TopicsListItem] =
    val partFilter = SQLSyntax.join(
      Seq(
        sqls"AND topics.groupid = ${group.id}",
        sqls"AND topics.sticky",
        tagId.map(t => sqls"AND topics.id IN (SELECT msgid FROM tags WHERE tagid=${t})").getOrElse(SQLSyntax.empty)
      ),
      sqls" "
    )

    load(
      partFilter = partFilter,
      topics = 100,
      offset = 0,
      sortByTopic = true,
      commentInterval = SQLSyntax.empty,
      topicInterval = SQLSyntax.empty,
      showIgnored = true,
      showDeleted = false,
      showUncommited = false
    )

  def getTrackerTopics(filter: TrackerFilterEnum, offset: Int)(using
      session: AnySession): collection.Seq[TopicsListItem] =
    val sectionFilter: SQLSyntax =
      filter match
        case TrackerFilterEnum.NOTALKS =>
          sqls"AND NOT topics.groupid = 8404 AND NOT topics.notop"
        case TrackerFilterEnum.MAIN =>
          sqls"AND NOT topics.groupid IN (${NonTech}) AND NOT topics.notop"
        case TrackerFilterEnum.TECH =>
          sqls"AND NOT topics.groupid IN (${NonTech}) AND NOT topics.notop AND section=2"
        case _ =>
          SQLSyntax.empty

    val partFilter = SQLSyntax.join(
      Seq(sectionFilter, sqls"AND lastmod > CURRENT_TIMESTAMP - '7 days'::interval"),
      sqls" ")

    load(
      partFilter = partFilter,
      topics = session.profile.topics,
      offset = offset,
      sortByTopic = false,
      commentInterval = sqls"AND comments.postdate > CURRENT_TIMESTAMP - '7 days'::interval",
      topicInterval = sqls"t.postdate > CURRENT_TIMESTAMP - '7 days'::interval",
      showIgnored = false,
      showDeleted = false,
      showUncommited = filter == TrackerFilterEnum.ALL || session.moderator || session.corrector
    )

  private def load(
      partFilter: SQLSyntax,
      topics: Int,
      offset: Int,
      sortByTopic: Boolean,
      commentInterval: SQLSyntax,
      topicInterval: SQLSyntax,
      showIgnored: Boolean,
      showDeleted: Boolean,
      showUncommited: Boolean)(using session: AnySession): collection.Seq[TopicsListItem] =
    val innerSortLimit: SQLSyntax =
      if sortByTopic then
        sqls"ORDER BY postdate DESC LIMIT $topics OFFSET $offset"
      else
        SQLSyntax.empty

    val outerSortLimit: SQLSyntax =
      if sortByTopic then
        sqls"ORDER BY topic_postdate DESC"
      else
        sqls"ORDER BY comment_postdate DESC LIMIT $topics OFFSET $offset"

    val userId =
      if session.authorized && !showIgnored then
        Some(session.userOpt.get.id)
      else
        None

    val partIgnored: SQLSyntax = userId.map(GroupListDao.queryPartIgnored).getOrElse(SQLSyntax.empty)
    val commentIgnored: SQLSyntax = userId.map(GroupListDao.queryPartCommentIgnored).getOrElse(SQLSyntax.empty)
    val tagIgnored: SQLSyntax = userId.map(GroupListDao.queryPartTagIgnored).getOrElse(SQLSyntax.empty)

    val noHidden: SQLSyntax =
      if session.authorized then
        SQLSyntax.empty
      else
        sqls"AND topics.open_warnings <= ${TopicMaxWarnings}"
    val partUncommited: SQLSyntax =
      if showUncommited then
        SQLSyntax.empty
      else
        GroupListDao.NoUncommited
    val partDeleted: SQLSyntax =
      if showDeleted then
        SQLSyntax.empty
      else
        sqls"AND NOT deleted"
    val whereTopicInterval: SQLSyntax =
      if topicInterval.isEmpty then
        SQLSyntax.empty
      else
        sqls"WHERE $topicInterval"

    springDB.run:
      sql"""WITH topics AS (
        SELECT topics.*, sections.moderate as smod, groups.title AS gtitle, urlname, section FROM topics
        JOIN groups ON topics.groupid=groups.id JOIN sections ON sections.id = groups.section
        WHERE NOT draft $partDeleted $partIgnored $tagIgnored $partUncommited $partFilter $noHidden $innerSortLimit
      ) SELECT * FROM (SELECT DISTINCT ON(id) * FROM (SELECT
        t.userid as author, t.id, t.stat1 AS stat1, gtitle, t.title AS title,
        comments.id as cid, comments.userid AS last_comment_by, t.resolved as resolved,
        section, urlname, comments.postdate as comment_postdate, smod, t.moderate,
        t.sticky, t.postdate as topic_postdate, t.deleted, t.postscore as topic_postscore
        FROM topics AS t JOIN comments ON (t.id=comments.topic)
        WHERE t.postscore IS DISTINCT FROM ${TopicPermissionService.POSTSCORE_HIDE_COMMENTS}
        AND comments.id=(SELECT id FROM comments WHERE NOT deleted AND comments.topic=t.id
        $commentIgnored ORDER BY postdate DESC LIMIT 1)
        $commentInterval
        UNION ALL
        SELECT
        t.userid as author, t.id, t.stat1 AS stat1, gtitle, t.title AS title,
        0, 0, t.resolved as resolved, section, urlname,
        postdate as comment_postdate, smod, t.moderate, t.sticky,
        t.postdate as topic_postdate, t.deleted, t.postscore as topic_postscore
        FROM topics AS t $whereTopicInterval) as tracker ORDER BY id, comment_postdate DESC) tracker $outerSortLimit"""
        .map(mapTopicsListItem)
        .list
        .apply()
        .toSeq

  private def mapTopicsListItem(rs: WrappedResultSet): TopicsListItem =
    TopicsListItem(
      topicAuthor = rs.int("author"),
      topicId = rs.int("id"),
      commentCount = rs.int("stat1"),
      groupTitle = rs.string("gtitle"),
      title = StringUtil.makeTitle(rs.string("title")),
      lastCommentId = rs.intOpt("cid").filter(_ != 0),
      lastCommentBy = rs.intOpt("last_comment_by").filter(_ != 0),
      resolved = rs.booleanOpt("resolved").getOrElse(false),
      section = rs.int("section"),
      groupUrlName = rs.string("urlname"),
      postdate = rs.timestamp("comment_postdate"),
      uncommited = rs.booleanOpt("smod").getOrElse(false) && !rs.booleanOpt("moderate").getOrElse(false),
      deleted = rs.booleanOpt("deleted").getOrElse(false),
      sticky = rs.booleanOpt("sticky").getOrElse(false),
      topicPostscore = rs.intOpt("topic_postscore").getOrElse(TopicPermissionService.POSTSCORE_UNRESTRICTED)
    )
